File: src/televault/preview.py
"""Preview engine for TeleVault - generate terminal previews without full download."""

import logging
import struct
from dataclasses import dataclass, field
from pathlib import Path

from .compress import decompress_data
from .config import Config
from .core import TeleVault
from .crypto import decrypt_chunk
from .models import FileMetadata
from .telegram import TelegramConfig
from .utils import format_size as _fmt_size

logger = logging.getLogger("televault.preview")

IMAGE_EXTENSIONS = {
    ".jpg",
    ".jpeg",
    ".png",
    ".gif",
    ".webp",
    ".bmp",
    ".ico",
    ".tiff",
    ".tif",
    ".heic",
    ".heif",
    ".avif",
}
VIDEO_EXTENSIONS = {
    ".mp4",
    ".mkv",
    ".avi",
    ".mov",
    ".webm",
    ".m4v",
    ".wmv",
    ".flv",
    ".mpg",
    ".mpeg",
    ".3gp",
}
AUDIO_EXTENSIONS = {
    ".mp3",
    ".wav",
    ".ogg",
    ".flac",
    ".aac",
    ".m4a",
    ".opus",
    ".wma",
    ".aiff",
}
TEXT_EXTENSIONS = {
    ".txt",
    ".md",
    ".py",
    ".js",
    ".ts",
    ".go",
    ".rs",
    ".c",
    ".cpp",
    ".h",
    ".java",
    ".rb",
    ".php",
    ".html",
    ".css",
    ".json",
    ".xml",
    ".yaml",
    ".yml",
    ".toml",
    ".ini",
    ".cfg",
    ".conf",
    ".sh",
    ".bash",
    ".zsh",
    ".fish",
    ".ps1",
    ".bat",
    ".sql",
    ".r",
    ".csv",
    ".log",
    ".env",
}
DOCUMENT_EXTENSIONS = {
    ".pdf",
    ".doc",
    ".docx",
    ".xls",
    ".xlsx",
    ".ppt",
    ".pptx",
    ".odt",
    ".ods",
    ".odp",
}
ARCHIVE_EXTENSIONS = {
    ".zip",
    ".tar",
    ".gz",
    ".bz2",
    ".xz",
    ".7z",
    ".rar",
    ".zst",
}

PREVIEW_SIZE_SMALL = 2048
PREVIEW_SIZE_MEDIUM = 65536
PREVIEW_SIZE_LARGE = 524288


@dataclass
class FilePreview:
    """Preview result for a file."""

    file_id: str
    name: str
    size: int
    file_type: str  # "image", "video", "audio", "text", "document", "archive", "binary"
    preview_text: str = ""
    metadata: dict = field(default_factory=dict)
    hex_dump: str = ""
    first_bytes: bytes = b""

    def to_dict(self) -> dict:
        d = {
            "file_id": self.file_id,
            "name": self.name,
            "size": self.size,
            "file_type": self.file_type,
            "metadata": self.metadata,
        }
        if self.preview_text:
            d["preview_text"] = self.preview_text
        if self.hex_dump:
            d["hex_dump"] = self.hex_dump
        return d


def classify_file(filename: str) -> str:
    """Classify a file by its extension."""
    ext = Path(filename).suffix.lower()
    if ext in IMAGE_EXTENSIONS:
        return "image"
    if ext in VIDEO_EXTENSIONS:
        return "video"
    if ext in AUDIO_EXTENSIONS:
        return "audio"
    if ext in TEXT_EXTENSIONS:
        return "text"
    if ext in DOCUMENT_EXTENSIONS:
        return "document"
    if ext in ARCHIVE_EXTENSIONS:
        return "archive"
    return "binary"


def generate_hex_dump(data: bytes, max_bytes: int = 256) -> str:
    """Generate a hex dump of the first N bytes."""
    lines = []
    data = data[:max_bytes]
    for offset in range(0, len(data), 16):
        chunk = data[offset : offset + 16]
        hex_part = " ".join(f"{b:02x}" for b in chunk)
        ascii_part = "".join(chr(b) if 32 <= b < 127 else "." for b in chunk)
        lines.append(f"{offset:08x}  {hex_part:<48s}  {ascii_part}")
    return "\n".join(lines)


def generate_text_preview(data: bytes, max_lines: int = 40) -> str:
    """Generate a text preview from raw bytes."""
    try:
        text = data.decode("utf-8", errors="replace")
    except Exception:
        text = data.decode("latin-1", errors="replace")

    lines = text.splitlines()
    if len(lines) > max_lines:
        return "\n".join(lines[:max_lines]) + f"\n... ({len(lines)} total lines)"
    return "\n".join(lines)


def extract_image_metadata(data: bytes) -> dict:
    """Extract metadata from image headers."""
    meta = {}

    if data[:8] == b"\x89PNG\r\n\x1a\n":
        meta["format"] = "PNG"
        if len(data) > 24:
            w = struct.unpack(">I", data[16:20])[0]
            h = struct.unpack(">I", data[20:24])[0]
            meta["width"] = w
            meta["height"] = h

    elif data[:2] == b"\xff\xd8":
        meta["format"] = "JPEG"
        i = 2
        while i < min(len(data) - 1, 65536):
            if data[i] != 0xFF:
                break
            marker = data[i + 1]
            if marker in (0xC0, 0xC1, 0xC2):
                if i + 9 <= len(data):
                    h = struct.unpack(">H", data[i + 5 : i + 7])[0]
                    w = struct.unpack(">H", data[i + 7 : i + 9])[0]
                    meta["width"] = w
                    meta["height"] = h
                break
            elif marker in (0xD0, 0xD1, 0xD2, 0xD3, 0xD4, 0xD5, 0xD6, 0xD7, 0xD8, 0xD9):
                i += 2
            elif i + 3 < len(data):
                length = struct.unpack(">H", data[i + 2 : i + 4])[0]
                i += 2 + length
            else:
                break

    elif data[:4] == b"RIFF" and data[8:12] == b"WEBP":
        meta["format"] = "WebP"

    elif data[:6] in (b"GIF87a", b"GIF89a"):
        meta["format"] = "GIF"
        if len(data) > 10:
            w = struct.unpack("<H", data[6:8])[0]
            h = struct.unpack("<H", data[8:10])[0]
            meta["width"] = w
            meta["height"] = h

    elif data[:4] == b"BM":
        meta["format"] = "BMP"

    return meta


def extract_video_metadata(data: bytes) -> dict:
    """Extract metadata from video file headers."""
    meta = {}

    if data[:4] == b"\x1a\x45\xdf\xa5":
        meta["format"] = "MKV/WebM"
    elif data[4:8] == b"ftyp":
        brand = data[8:12].decode("ascii", errors="replace").strip()
        meta["format"] = f"MP4 ({brand})"
    elif data[:4] == b"RIFF" and len(data) > 10 and data[8:12] == b"AVI ":
        meta["format"] = "AVI"

    return meta


def extract_audio_metadata(data: bytes) -> dict:
    """Extract metadata from audio file headers."""
    meta = {}

    if data[:3] == b"ID3":
        meta["format"] = "MP3 (ID3)"
        try:
            size_flags = data[6:10]
            synsafe = 0
            for b in size_flags:
                synsafe = synsafe * 128 + b
            id3_end = 10 + synsafe
            meta["id3_size"] = id3_end
        except Exception:
            pass

    elif data[:4] == b"fLaC":
        meta["format"] = "FLAC"

    elif data[:4] == b"OggS":
        meta["format"] = "OGG"

    elif data[:4] == b"RIFF" and data[8:12] == b"WAVE":
        meta["format"] = "WAV"
        if len(data) > 28:
            channels = struct.unpack("<H", data[22:24])[0]
            sample_rate = struct.unpack("<I", data[24:28])[0]
            meta["channels"] = channels
            meta["sample_rate"] = sample_rate

    return meta


def render_ascii_image(width: int, height: int, format_name: str, size: int) -> str:
    """Render an ASCII representation of an image file."""
    aspect = width / height if height > 0 else 1
    cols = min(40, max(20, int(aspect * 10)))
    rows = max(5, min(12, int(cols / aspect / 2)))

    block = "\u2591"
    lines = []
    lines.append(f"  +{'─' * cols}+")
    for _ in range(rows):
        lines.append(f"  |{block * cols}|")
    lines.append(f"  +{'─' * cols}+")

    dim_line = f"  {width}x{height} {format_name}" if width and height else f"  {format_name}"
    lines.insert(1, dim_line)

    return "\n".join(lines)


class PreviewEngine:
    """Generate terminal previews for vault files without full download."""

    def __init__(
        self,
        config: Config | None = None,
        telegram_config: TelegramConfig | None = None,
        password: str | None = None,
    ):
        self.config = config or Config.load_or_create()
        self.password = password
        self._telegram_config = telegram_config
        self._vault: TeleVault | None = None

    async def _ensure_connected(self):
        if self._vault is None:
            self._vault = TeleVault(
                config=self.config,
                telegram_config=self._telegram_config,
                password=self.password,
            )
            await self._vault.connect()

    async def disconnect(self):
        if self._vault:
            await self._vault.disconnect()
            self._vault = None

    async def preview(
        self,
        file_id_or_name: str,
        size: str = "small",
        password: str | None = None,
    ) -> FilePreview:
        """Generate a preview for a file in the vault."""
        await self._ensure_connected()
        password = password or self.password

        index = await self._vault.telegram.get_index()

        metadata_msg_id = None
        if file_id_or_name in index.files:
            metadata_msg_id = index.files[file_id_or_name]
        else:
            files = await self._vault.list_files()
            matches = [f for f in files if f.name == file_id_or_name or file_id_or_name in f.name]
            if not matches:
                raise FileNotFoundError(f"File not found: {file_id_or_name}")
            if len(matches) > 1:
                raise ValueError(f"Multiple files match '{file_id_or_name}'")
            metadata_msg_id = matches[0].message_id

        metadata = await self._vault.telegram.get_metadata(metadata_msg_id)
        file_type = classify_file(metadata.name)

        size_bytes = {
            "small": PREVIEW_SIZE_SMALL,
            "medium": PREVIEW_SIZE_MEDIUM,
            "large": PREVIEW_SIZE_LARGE,
        }.get(size, PREVIEW_SIZE_SMALL)

        first_bytes = await self._download_first_bytes(metadata, size_bytes, password)

        preview_text = ""
        file_metadata = {}

        if file_type == "image":
            file_metadata = extract_image_metadata(first_bytes)
            w = file_metadata.get("width", 0)
            h = file_metadata.get("height", 0)
            if w and h:
                file_metadata["dimensions"] = f"{w}x{h}"
            file_metadata["file_size"] = _fmt_size(metadata.size)
            if metadata.compressed:
                file_metadata["compression"] = "Yes"
            if metadata.encrypted:
                file_metadata["encryption"] = "Yes"
            preview_text = ""

        elif file_type == "video":
            file_metadata = extract_video_metadata(first_bytes)
            file_metadata["file_size"] = _fmt_size(metadata.size)
            if metadata.compressed:
                file_metadata["compression"] = "Yes"
            if metadata.encrypted:
                file_metadata["encryption"] = "Yes"
            preview_text = ""

        elif file_type == "audio":
            file_metadata = extract_audio_metadata(first_bytes)
            file_metadata["file_size"] = _fmt_size(metadata.size)
            if metadata.compressed:
                file_metadata["compression"] = "Yes"
            if metadata.encrypted:
                file_metadata["encryption"] = "Yes"
            preview_text = ""

        elif file_type == "text":
            preview_text = generate_text_preview(first_bytes, max_lines=40)

        else:
            hex_dump = generate_hex_dump(first_bytes)
            preview_text = hex_dump

        return FilePreview(
            file_id=metadata.id,
            name=metadata.name,
            size=metadata.size,
            file_type=file_type,
            preview_text=preview_text,
            metadata=file_metadata,
            hex_dump=generate_hex_dump(first_bytes) if file_type not in ("text",) else "",
            first_bytes=first_bytes[:256],
        )

    async def _download_first_bytes(
        self, metadata: FileMetadata, max_bytes: int, password: str | None
    ) -> bytes:
        """Download just enough chunks to get the first max_bytes."""
        if not metadata.chunks:
            return b""

        sorted_chunks = sorted(metadata.chunks, key=lambda c: c.index)
        first_chunk = sorted_chunks[0]

        data = await self._vault.telegram.download_chunk(first_chunk.message_id)

        if metadata.encrypted and password:
            data = decrypt_chunk(data, password)

        if metadata.compressed:
            data = decompress_data(data)

        return data[:max_bytes]

File: src/preview/preview.cpp
#include "preview.hpp"
#include <spdlog/spdlog.h>
#include <array>
#include <unordered_map>
#include <cctype>
#include <fstream>
#include <sstream>
#include <iomanip>
#include <algorithm>
#include <ranges>
#include <filesystem>

namespace tv {

static const std::unordered_map<std::string, FileCategory> ext_category = {
    {".txt", FileCategory::Text},
    {".md", FileCategory::Text},
    {".json", FileCategory::Text},
    {".xml", FileCategory::Text},
    {".yaml", FileCategory::Text},
    {".yml", FileCategory::Text},
    {".csv", FileCategory::Text},
    {".log", FileCategory::Text},
    {".py", FileCategory::Text},
    {".cpp", FileCategory::Text},
    {".c", FileCategory::Text},
    {".h", FileCategory::Text},
    {".hpp", FileCategory::Text},
    {".js", FileCategory::Text},
    {".ts", FileCategory::Text},
    {".html", FileCategory::Text},
    {".css", FileCategory::Text},
    {".sh", FileCategory::Text},
    {".rs", FileCategory::Text},
    {".go", FileCategory::Text},
    {".toml", FileCategory::Text},
    {".ini", FileCategory::Text},
    {".cfg", FileCategory::Text},

    {".jpg", FileCategory::Image},
    {".jpeg", FileCategory::Image},
    {".png", FileCategory::Image},
    {".gif", FileCategory::Image},
    {".webp", FileCategory::Image},
    {".bmp", FileCategory::Image},
    {".svg", FileCategory::Image},

    {".mp4", FileCategory::Video},
    {".mkv", FileCategory::Video},
    {".avi", FileCategory::Video},
    {".mov", FileCategory::Video},
    {".webm", FileCategory::Video},

    {".mp3", FileCategory::Audio},
    {".m4a", FileCategory::Audio},
    {".ogg", FileCategory::Audio},
    {".opus", FileCategory::Audio},
    {".flac", FileCategory::Audio},
    {".wav", FileCategory::Audio},
    {".aac", FileCategory::Audio},
};

static const std::unordered_map<std::string, std::string> ext_mime = {
    {".txt", "text/plain"},
    {".md", "text/markdown"},
    {".json", "application/json"},
    {".xml", "application/xml"},
    {".csv", "text/csv"},
    {".html", "text/html"},
    {".css", "text/css"},
    {".js", "application/javascript"},
    {".jpg", "image/jpeg"},
    {".jpeg", "image/jpeg"},
    {".png", "image/png"},
    {".gif", "image/gif"},
    {".webp", "image/webp"},
    {".svg", "image/svg+xml"},
    {".mp4", "video/mp4"},
    {".mp3", "audio/mpeg"},
    {".pdf", "application/pdf"},
    {".zip", "application/zip"},
};

class PreviewEngine::Impl {
public:
    FileCategory classify(const std::string& filename) const {
        auto pos = filename.rfind('.');
        if (pos == std::string::npos) return FileCategory::Binary;

        auto ext = filename.substr(pos);
        std::string lower;
        lower.resize(ext.size());
        std::ranges::transform(ext, lower.begin(), [](unsigned char c) {
            return static_cast<char>(std::tolower(c));
        });

        auto it = ext_category.find(lower);
        return it != ext_category.end() ? it->second : FileCategory::Binary;
    }

    std::string get_mime(const std::string& filename) const {
        auto pos = filename.rfind('.');
        if (pos == std::string::npos) return "application/octet-stream";

        auto ext = filename.substr(pos);
        std::string lower;
        lower.resize(ext.size());
        std::ranges::transform(ext, lower.begin(), [](unsigned char c) {
            return static_cast<char>(std::tolower(c));
        });

        auto it = ext_mime.find(lower);
        return it != ext_mime.end() ? it->second : "application/octet-stream";
    }

    PreviewResult preview_file(const std::string& local_path, const std::string& original_filename = "") {
        PreviewResult result;
        std::string name_for_type = original_filename.empty() ? local_path : original_filename;
        result.category = classify(name_for_type);
        result.mime_type = get_mime(name_for_type);

        switch (result.category) {
            case FileCategory::Text:
                result.text_preview = preview_text(local_path);
                break;
            case FileCategory::Image:
                result.text_preview = preview_image(local_path);
                break;
            default:
                result.text_preview = preview_hex(local_path);
                break;
        }

        return result;
    }

private:
    std::string preview_text(const std::string& path, int max_lines = 30) {
        std::ifstream f(path);
        if (!f.is_open()) return "(cannot open)";

        std::ostringstream out;
        std::string line;
        int count = 0;
        while (std::getline(f, line) && count < max_lines) {
            out << line << "\n";
            ++count;
        }
        if (f.good()) out << "... (" << max_lines << " lines shown)\n";
        return out.str();
    }

    std::string preview_image(const std::string& path) {
        // Read image headers for basic metadata
        std::ifstream f(path, std::ios::binary);
        if (!f.is_open()) return "(cannot open)";

        auto size = std::filesystem::file_size(path);
        std::ostringstream out;
        out << "Image file\n";
        out << "Size: " << size << " bytes\n";

        // Check for PNG/JPEG headers
        std::array<char, 8> header{};
        f.read(header.data(), header.size());

        if (header[0] == char(0x89) && header[1] == 'P' && header[2] == 'N' && header[3] == 'G') {
            out << "Format: PNG\n";
        } else if (header[0] == char(0xFF) && header[1] == char(0xD8)) {
            out << "Format: JPEG\n";
        } else if (header[0] == 'G' && header[1] == 'I' && header[2] == 'F') {
            out << "Format: GIF\n";
        } else if (header[0] == 'R' && header[1] == 'I' && header[2] == 'F' && header[3] == 'F') {
            out << "Format: WebP\n";
        }

        return out.str();
    }

    std::string preview_hex(const std::string& path, int max_bytes = 256) {
        std::ifstream f(path, std::ios::binary);
        if (!f.is_open()) return "(cannot open)";

        std::vector<char> buf(max_bytes);
        auto read = f.read(buf.data(), buf.size()).gcount();

        std::ostringstream out;
        out << "Binary file (" << std::filesystem::file_size(path) << " bytes)\n\n";
        out << "Hex dump (first " << read << " bytes):\n";

        for (std::streamsize i = 0; i < read; i += 16) {
            auto remaining = std::min<std::streamsize>(16, read - i);
            out << std::hex << std::setfill('0') << std::setw(8) << i << "  ";

            for (std::streamsize j = 0; j < 16; ++j) {
                if (j < remaining) {
                    out << std::hex << std::setfill('0') << std::setw(2)
                        << (static_cast<int>(buf[i + j]) & 0xFF) << " ";
                } else {
                    out << "   ";
                }
                if (j == 7) out << " ";
            }

            out << " |";
            for (std::streamsize j = 0; j < remaining; ++j) {
                auto c = static_cast<unsigned char>(buf[i + j]);
                out << (std::isprint(c) ? static_cast<char>(c) : '.');
            }
            out << "|\n";
        }

        return out.str();
    }
};

PreviewEngine::PreviewEngine() : impl_(std::make_unique<Impl>()) {}
PreviewEngine::~PreviewEngine() = default;

FileCategory PreviewEngine::classify(const std::string& filename) const {
    return impl_->classify(filename);
}

std::string PreviewEngine::mime_type(const std::string& filename) const {
    return impl_->get_mime(filename);
}

PreviewResult PreviewEngine::preview(const std::string& vault_path, const std::string& original_filename) {
    return impl_->preview_file(vault_path, original_filename);
}

} // namespace tv
